From 51b5b6c5afbbfb2dd5f626ce854c6a6d03f27a31 Mon Sep 17 00:00:00 2001
From: Rajesh Balamohan <rbalamohan at apache dot org>
Date: Thu, 21 Jul 2016 08:01:33 -0700
Subject: [PATCH 0823/1164] CDH-47359: HIVE-14301 : insert overwrite fails for
 nonpartitioned tables in s3 (Rajesh Balamohan via
 Ashutosh Chauhan)

Signed-off-by: Ashutosh Chauhan <hashutosh@apache.org>
(cherry picked from commit 8c546fe337d7900e4b5fa5b308ce5b49547e0f4b)

Conflicts:
	common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
	ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

Change-Id: Id77166c1e7e34b1736c8a39a184a918f2b837307
---
 .../java/org/apache/hadoop/hive/conf/HiveConf.java |    2 +-
 .../org/apache/hadoop/hive/ql/metadata/Hive.java   |   10 +++++-----
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java b/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
index 35406ec..bec1c53 100644
--- a/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
+++ b/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
@@ -1949,7 +1949,7 @@ public void setSparkConfigUpdated(boolean isSparkConfigUpdated) {
         "If set, this configuration property should provide a comma-separated list of URLs that indicates the type and " +
         "location of providers to be used by hadoop credential provider API. It provides HiveServer2 the ability to provide job-specific " +
         "credential providers for jobs run using MR and Spark execution engines. This functionality has not been tested against Tez."),
-    HIVE_MOVE_FILES_THREAD_COUNT("hive.mv.files.thread", 25, "Number of threads"
+    HIVE_MOVE_FILES_THREAD_COUNT("hive.mv.files.thread", 15, "Number of threads"
          + " used to move files in move task. Set it to 0 to disable multi-threaded file moves. This parameter is also used by"
          + " MSCK to check tables."),
     // If this is set all move tasks at the end of a multi-insert query will only begin once all
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
index 3e48959..cf894f3 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
@@ -2785,8 +2785,9 @@ public static boolean moveFile(final HiveConf conf, Path srcf, final Path destf,
             /* Move files one by one because source is a subdirectory of destination */
             for (final FileStatus srcStatus : srcs) {
 
+              final Path destFile = new Path(destf, srcStatus.getPath().getName());
               if (null == pool) {
-                if(!destFs.rename(srcStatus.getPath(), destf)) {
+                if(!destFs.rename(srcStatus.getPath(), destFile)) {
                   throw new IOException("rename for src path: " + srcStatus.getPath() + " to dest:"
                       + destf + " returned false");
                 }
@@ -2795,15 +2796,14 @@ public static boolean moveFile(final HiveConf conf, Path srcf, final Path destf,
                   @Override
                   public Void call() throws Exception {
                     SessionState.setCurrentSessionState(parentSession);
-                    final Path destPath = new Path(destf, srcStatus.getPath().getName());
                     final String group = srcStatus.getGroup();
-                    if(destFs.rename(srcStatus.getPath(), destf)) {
+                    if(destFs.rename(srcStatus.getPath(), destFile)) {
                       if (inheritPerms) {
-                        ShimLoader.getHadoopShims().setFullFileStatus(conf, desiredStatus, null, destFs, destPath, false);
+                        ShimLoader.getHadoopShims().setFullFileStatus(conf, desiredStatus, group, destFs, destFile, false);
                       }
                     } else {
                       throw new IOException("rename for src path: " + srcStatus.getPath() + " to dest path:"
-                          + destPath + " returned false");
+                          + destFile + " returned false");
                     }
                     return null;
                   }
-- 
1.7.9.5

